File: src/polls/ingest/wiki_polls.py
"""SUPERSEDED for the live loop (2026-09-02): the production watcher is the
Deno port, js/ingest/wiki_polls.ts (parity was proven on identical HTML before
the switch, then the harness deleted). This module remains ONLY because the
archival full-rebuild (ingest/run.py) and the next-election recalibration
(model/regional_calibration.py) import it. Do not fix parser bugs here first —
fix them in the TS parser, which is what actually feeds the dashboard.

Poll ingestion from Wikipedia's maintained polling tables.

Replaces the qc125.com scrape as the source of vote-intention polls. Two
reasons, and the second is the one that matters:

  - **Provenance.** qc125 is a competing projection site that compiles polls
    as part of its own product. Taking its compilation to build a rival
    projection, and then not saying so, would be the worst of both worlds.
    Wikipedia's tables are CC BY-SA, explicitly meant for reuse, and each row
    carries a link to the pollster's own release -- so `source_url` points at
    the primary document rather than at an intermediary's copy.
  - **Coverage is better.** Measured, not assumed. Across the three cycle
    pages this yields 538 provincial polls spanning 2014-2026, against 195 in
    the qc125-derived database. For the current cycle the two sources carry
    essentially the same polls: a side-by-side match on firm and sample size
    shows them pairing one-for-one with dates offset by exactly one day (the
    two sites use different date conventions -- last field day vs. release).
    Each source has a handful the other lacks; Wikipedia was also one day
    fresher at the time of the switch.

Because of that one-day offset the poll_id hashes do NOT line up between the
two sources, so this is a replacement rather than a merge -- running both
would double-count nearly every poll. The id scheme is kept identical anyway
so that re-ingesting the same Wikipedia page twice is idempotent.

What is lost: qc125's regional subsamples (Montreal / Quebec City / rest of
province) and its editorial pollster ratings. Neither is used by the live
model -- only the National series is exported, and the rating's measured
contribution to aggregation accuracy was 0.01pp. A missing rating falls back
to a constant weight, which after normalisation is the same as no rating
weighting at all.

Column layout differs between cycles (2018 lists PQ before QS and carries a
PVQ column that later years drop), so parsing is driven by column NAMES, never
by position.

"""

from __future__ import annotations

import hashlib
import re
import unicodedata
from dataclasses import dataclass
from datetime import date, datetime

import httpx
import lxml.html
import polars as pl

USER_AGENT = "polls-research-bot/0.1 (contact: yysrk9lh@duck.com; Quebec election forecasting project)"
API = "https://fr.wikipedia.org/w/api.php"

POLL_PAGES: dict[str, str] = {
    "2026": "Liste de sondages sur les élections générales québécoises de 2026",
    "2022": "Liste de sondages sur les élections générales québécoises de 2022",
    "2018": "Liste de sondages sur les élections générales québécoises de 2018",
}

# Wikipedia's column headers -> our party codes. PVQ (Green) appears only in
# the 2018 table and is folded into AUTRES rather than given a code of its
# own: it is absent from every other cycle, and a party present in one
# election and structurally missing in the next cannot be carried as its own
# part of the composition.
PARTY_COLUMNS = {
    "CAQ": "CAQ",
    "PLQ": "LIB",
    "QS": "QS",
    "PQ": "PQ",
    "PCQ": "PCQ",
    "CQ": "PCQ",
    "PVQ": "AUTRES",
    "Autres": "AUTRES",
}

# Wikipedia abbreviates; the database already holds these spellings from the
# previous source. Normalising here keeps poll_id stable across the switch.
FIRM_NORMALIZE = {
    "Mainstreet": "Mainstreet Research",
    "Pallas": "Pallas Data",
    "Liaison": "Liaison Strategies",
    "Innovative": "Innovative Research",
    "Synopsis": "Synopsis Recherche",
    "Synopsis / La Presse": "Synopsis Recherche",
    "Segma": "Segma Recherche",
    "Research Co": "Research Co.",
    "Forum": "Forum Research",
}

MONTHS = {
    "janvier": 1, "février": 2, "fevrier": 2, "mars": 3, "avril": 4, "mai": 5,
    "juin": 6, "juillet": 7, "août": 8, "aout": 8, "septembre": 9,
    "octobre": 10, "novembre": 11, "décembre": 12, "decembre": 12,
}


@dataclass(frozen=True)
class PageSnapshot:
    """A fetched page plus the revision id that identifies it. The revision id
    is what makes a cheap watch possible: one small API call tells you whether
    anything changed at all, without re-downloading or re-parsing 380 KB."""

    title: str
    revid: int
    html: str


def latest_revision(title: str) -> int:
    resp = httpx.get(
        API,
        params={"action": "query", "prop": "revisions", "titles": title, "rvprop": "ids", "format": "json"},
        headers={"User-Agent": USER_AGENT},
        timeout=30,
    )
    resp.raise_for_status()
    pages = resp.json()["query"]["pages"]
    page = next(iter(pages.values()))
    if "revisions" not in page:
        raise LookupError(f"page introuvable sur fr.wikipedia : {title}")
    return int(page["revisions"][0]["revid"])


def fetch_page(title: str) -> PageSnapshot:
    # Revision id BEFORE the HTML, not after. The other order has a race: an
    # edit landing between the two requests records a revid whose content was
    # never parsed, and any poll in that edit is silently missed until the
    # page is edited again. This order errs the safe way -- an edit in the gap
    # makes the stored revid stale, so the next watch run re-examines it.
    revid = latest_revision(title)
    url = f"https://fr.wikipedia.org/wiki/{title.replace(' ', '_')}"
    resp = httpx.get(url, headers={"User-Agent": USER_AGENT}, timeout=60, follow_redirects=True)
    resp.raise_for_status()
    # Decode explicitly: httpx's charset sniffing has mis-guessed on these
    # pages before, and a mis-decoded "août" silently kills date parsing.
    return PageSnapshot(title=title, revid=revid, html=resp.content.decode("utf-8"))


def _parse_date(text: str) -> date | None:
    text = re.sub(r"\[.*?\]", "", text).strip().lower().replace("\xa0", " ")
    m = re.search(r"(\d{1,2})\s+([a-zéûôà]+)\s+(\d{4})", text)
    if not m:
        return None
    month = MONTHS.get(m.group(2))
    return date(int(m.group(3)), month, int(m.group(1))) if month else None


def _parse_number(text: str) -> float | None:
    """Cells carry footnote markers, thin spaces, commas as decimal separators
    and occasional en-dashes for "not asked"."""
    text = re.sub(r"\[.*?\]", "", str(text)).replace("\xa0", "").replace(" ", "").replace(" ", "")
    text = text.replace(",", ".").replace("%", "").replace("±", "")
    m = re.search(r"-?\d+(?:\.\d+)?", text)
    return float(m.group()) if m else None


def _normalize_firm(text: str) -> str:
    text = re.sub(r"\[.*?\]", "", text).strip()
    text = re.sub(r"\s*/\s*La Presse$", "", text)
    return FIRM_NORMALIZE.get(text, text)


def _poll_id(jurisdiction_code: str, region_code: str, firm: str, poll_date: str, sample: str | None) -> str:
    key = f"{jurisdiction_code}|{region_code}|{firm}|{poll_date}|{sample or ''}"
    return hashlib.sha1(key.encode("utf-8")).hexdigest()[:16]


# The cycle pages do not only carry province-wide polls. They also publish
# the same pollsters' SUBGROUP breakdowns -- by language, by region, by age --
# in tables with identical columns. Reading those as province-wide numbers put
# the Liberals at 80% in 2015 (that is the non-francophone subsample) and
# inflated the series with rows that are not what they claim to be.
#
# The subgroup is never in the table; it is in the section heading above it.
# So tables are located by walking the document in order and keeping the
# heading chain, then rejecting any table sitting under a breakdown heading.
BREAKDOWN_PATTERNS = (
    "par langue",
    "par region",
    "par age",
    "par sexe",
    "par groupe",
    "francophone",
    "chez les",
    "ile de montreal",
    "region metropolitaine",
    "capitale-nationale",
    "reste du quebec",
)


def _strip_accents(text: str) -> str:
    return unicodedata.normalize("NFKD", text).encode("ascii", "ignore").decode("ascii")


def _is_breakdown(headings: list[str]) -> bool:
    joined = _strip_accents(" / ".join(headings)).lower()
    return any(pattern in joined for pattern in BREAKDOWN_PATTERNS)


def _tables_with_headings(doc):
    """Yield (table, heading_chain) in document order, the chain being the
    current h2/h3/h4 above the table."""
    chain: dict[int, str] = {}
    for node in doc.xpath("//h2|//h3|//h4|//table"):
        tag = node.tag.lower()
        if tag in ("h2", "h3", "h4"):
            level = int(tag[1])
            chain[level] = _clean_heading(node.text_content())
            for deeper in [k for k in chain if k > level]:
                del chain[deeper]
        else:
            yield node, [chain[k] for k in sorted(chain)]


def _clean_heading(text: str) -> str:
    return re.sub(r"\s+", " ", text.replace("[modifier | modifier le code]", "")).strip()


def _cell_text(cell) -> str:
    return cell.text_content().strip()


def _cell_link(cell) -> str | None:
    """First external link in the cell -- the pollster's own release."""
    for href in cell.xpath(".//a/@href"):
        if href.startswith("http") and "wikipedia.org" not in href and "wikimedia.org" not in href:
            return href
    return None


def parse_polls(snapshot: PageSnapshot, jurisdiction_code: str = "qc-provincial") -> tuple[pl.DataFrame, pl.DataFrame]:
    """Returns (polls, poll_shares). Rows that are not polls -- the tables
    interleave narrative event rows like "14 janvier 2026: Annonce de la
    démission de François Legault" -- are dropped by requiring both a
    parseable date and at least two numeric party cells."""
    doc = lxml.html.fromstring(snapshot.html)
    poll_rows: list[dict] = []
    share_rows: list[dict] = []
    seen: set[str] = set()
    skipped: list[tuple[str, int]] = []

    for table, headings in _tables_with_headings(doc):
        headers = [_cell_text(th) for th in table.xpath(".//tr[1]/th")]
        if "Sondeur" not in headers or "CAQ" not in headers:
            continue
        if _is_breakdown(headings):
            skipped.append((headings[-1] if headings else "?", len(table.xpath(".//tr")) - 1))
            continue
        idx = {name: i for i, name in enumerate(headers)}
        party_idx = {headers[i]: i for i in range(len(headers)) if headers[i] in PARTY_COLUMNS}

        for tr in table.xpath(".//tr"):
            cells = tr.xpath("./td")
            if len(cells) < len(headers) - 2:
                continue

            poll_date = _parse_date(_cell_text(cells[0]))
            if poll_date is None:
                continue

            shares: dict[str, float] = {}
            for col, i in party_idx.items():
                if i >= len(cells):
                    continue
                value = _parse_number(_cell_text(cells[i]))
                if value is None:
                    continue
                code = PARTY_COLUMNS[col]
                shares[code] = shares.get(code, 0.0) + value
            if len(shares) < 2:
                continue

            firm_i = idx["Sondeur"]
            if firm_i >= len(cells):
                continue
            firm = _normalize_firm(_cell_text(cells[firm_i]))
            if not firm:
                continue

            sample = None
            if "Échantillon" in idx and idx["Échantillon"] < len(cells):
                raw = _parse_number(_cell_text(cells[idx["Échantillon"]]))
                sample = int(raw) if raw and raw > 0 else None

            source_url = None
            if "Source" in idx and idx["Source"] < len(cells):
                source_url = _cell_link(cells[idx["Source"]])

            pid = _poll_id(jurisdiction_code, "National", firm, poll_date.isoformat(), str(sample) if sample else None)
            if pid in seen:
                continue
            seen.add(pid)

            poll_rows.append(
                {
                    "poll_id": pid,
                    "jurisdiction_code": jurisdiction_code,
                    "region_code": "National",
                    "firm": firm,
                    "poll_date": poll_date,
                    "sample_size": sample,
                    "is_rolling": False,
                    "firm_rating": None,
                    "general_election": None,
                    "source_url": source_url or f"https://fr.wikipedia.org/wiki/{snapshot.title.replace(' ', '_')}",
                    "scraped_at": datetime.now(),
                }
            )
            for code, pct in shares.items():
                share_rows.append({"poll_id": pid, "party_code": code, "pct_reported": pct})

    if skipped:
        # Announce what was dropped: a silent filter looks identical to a
        # source that never had those tables.
        detail = ", ".join(f"{name} ({n})" for name, n in skipped)
        print(f"    sous-groupes ignores : {detail}")

    polls = pl.DataFrame(poll_rows, infer_schema_length=None) if poll_rows else pl.DataFrame()
    shares = pl.DataFrame(share_rows, infer_schema_length=None) if share_rows else pl.DataFrame()
    return polls, shares


def fetch_all(cycles: list[str] | None = None) -> tuple[pl.DataFrame, pl.DataFrame, dict[str, int]]:
    """All configured cycles, concatenated, plus the revision id seen for
    each page so a watcher can record what it has already processed."""
    frames_p, frames_s, revs = [], [], {}
    for cycle in cycles or list(POLL_PAGES):
        snapshot = fetch_page(POLL_PAGES[cycle])
        polls, shares = parse_polls(snapshot)
        revs[cycle] = snapshot.revid
        if len(polls):
            frames_p.append(polls)
            frames_s.append(shares)

    polls = pl.concat(frames_p, how="diagonal").unique(subset=["poll_id"], keep="first") if frames_p else pl.DataFrame()
    shares = pl.concat(frames_s, how="diagonal").unique(subset=["poll_id", "party_code"]) if frames_s else pl.DataFrame()
    return polls, shares, revs


REVISION_TABLE = "source_revisions"


def record_revisions(con, revisions: dict[str, int]) -> None:
    con.execute(
        f"CREATE TABLE IF NOT EXISTS {REVISION_TABLE} "
        "(source VARCHAR PRIMARY KEY, revid BIGINT, checked_at TIMESTAMP)"
    )
    for cycle, revid in revisions.items():
        con.execute(f"DELETE FROM {REVISION_TABLE} WHERE source = ?", [f"wiki_qc_{cycle}"])
        con.execute(f"INSERT INTO {REVISION_TABLE} VALUES (?, ?, now())", [f"wiki_qc_{cycle}", revid])


def known_revisions(con) -> dict[str, int]:
    """Revision ids already processed, keyed by cycle. Empty when the table
    doesn't exist yet, which makes the first watch run a full ingest."""
    try:
        rows = con.execute(f"SELECT source, revid FROM {REVISION_TABLE}").fetchall()
    except Exception:
        return {}
    return {s.removeprefix("wiki_qc_"): int(r) for s, r in rows if s.startswith("wiki_qc_")}


def ingest_qc_polls(con, cycles: list[str] | None = None) -> tuple[int, dict[str, int]]:
    """Fetch, load and record revisions. Replaces the provincial poll series
    wholesale: the previous source dated polls one day differently, so leaving
    its rows in place would double-count almost every poll."""
    from polls.db.load import upsert_jurisdiction, upsert_parties, upsert_poll_shares, upsert_polls, upsert_regions

    polls, shares, revisions = fetch_all(cycles)

    upsert_jurisdiction(con, "qc-provincial", "Québec (provincial)", "fr.wikipedia.org")
    upsert_parties(con, "qc-provincial", ["CAQ", "LIB", "QS", "PQ", "PCQ", "AUTRES"])
    upsert_regions(con, "qc-provincial", ["National"])

    # Replace ONLY the National series this source provides. Regional rows
    # (MTL/QC/REG) come from other loaders -- Léger's regional cumulation,
    # report PDFs -- and a blanket jurisdiction-wide delete silently destroyed
    # them on every watch run.
    con.execute(
        "DELETE FROM poll_shares WHERE poll_id IN "
        "(SELECT poll_id FROM polls WHERE jurisdiction_code = 'qc-provincial' AND region_code = 'National')"
    )
    con.execute("DELETE FROM polls WHERE jurisdiction_code = 'qc-provincial' AND region_code = 'National'")

    upsert_polls(con, polls)
    upsert_poll_shares(con, shares)
    record_revisions(con, revisions)
    return len(polls), revisions


if __name__ == "__main__":
    polls, shares, revs = fetch_all()
    print(f"{len(polls)} sondages, {len(shares)} parts, revisions {revs}")
    print(polls.sort("poll_date", descending=True).head(8).to_pandas().to_string())

File: js/ingest/wiki_polls.ts
/**
 * Poll ingestion from Wikipedia's maintained polling tables — Deno/TypeScript
 * port of the retired Python parser (tag `python-archive`), byte-for-byte on poll_id.
 *
 * The Python original carries the full rationale (provenance, coverage,
 * the one-day date-convention offset that makes this a replacement rather
 * than a merge). Two invariants matter most here and are enforced by the
 * parity test (ingest/parity.ts):
 *
 *   - poll_id = sha1("jurisdiction|region|firm|date|sample")[:16], so ids
 *     stay stable across the Python -> Deno switch and re-ingest stays
 *     idempotent.
 *   - Subgroup tables (by language, region, age...) share their column
 *     layout with the province-wide tables; only the SECTION HEADING above
 *     distinguishes them. Tables are walked in document order with their
 *     h2/h3/h4 chain, and any table under a breakdown heading is rejected.
 *     Reading those as province-wide once put the Liberals at 80% in 2015.
 */

import { createHash } from "node:crypto";
import { parseHTML } from "linkedom";

export const USER_AGENT =
  "polls-research-bot/0.1 (contact: yysrk9lh@duck.com; Quebec election forecasting project)";
const API = "https://fr.wikipedia.org/w/api.php";

export const POLL_PAGES: Record<string, string> = {
  "2026": "Liste de sondages sur les élections générales québécoises de 2026",
  "2022": "Liste de sondages sur les élections générales québécoises de 2022",
  "2018": "Liste de sondages sur les élections générales québécoises de 2018",
};

// Wikipedia's column headers -> our party codes. PVQ (Green) exists only in
// the 2018 table and folds into AUTRES: a party present in one election and
// structurally missing in the next cannot be its own part of the composition.
const PARTY_COLUMNS: Record<string, string> = {
  CAQ: "CAQ",
  PLQ: "LIB",
  QS: "QS",
  PQ: "PQ",
  PCQ: "PCQ",
  CQ: "PCQ",
  PVQ: "AUTRES",
  Autres: "AUTRES",
};

// Wikipedia abbreviates; the database already holds these spellings from the
// previous source. Normalising keeps poll_id stable across the switch.
const FIRM_NORMALIZE: Record<string, string> = {
  Mainstreet: "Mainstreet Research",
  Pallas: "Pallas Data",
  Liaison: "Liaison Strategies",
  Innovative: "Innovative Research",
  Synopsis: "Synopsis Recherche",
  "Synopsis / La Presse": "Synopsis Recherche",
  Segma: "Segma Recherche",
  "Research Co": "Research Co.",
  Forum: "Forum Research",
};

const MONTHS: Record<string, number> = {
  janvier: 1, "février": 2, fevrier: 2, mars: 3, avril: 4, mai: 5,
  juin: 6, juillet: 7, "août": 8, aout: 8, septembre: 9,
  octobre: 10, novembre: 11, "décembre": 12, decembre: 12,
};

export interface PageSnapshot {
  title: string;
  revid: number;
  html: string;
}

export interface PollRow {
  poll_id: string;
  jurisdiction_code: string;
  region_code: string;
  firm: string;
  poll_date: string; // ISO, kept as string end to end (no Date tz pitfalls)
  sample_size: number | null;
  is_rolling: boolean;
  firm_rating: null;
  general_election: null;
  source_url: string;
}

export interface ShareRow {
  poll_id: string;
  party_code: string;
  pct_reported: number;
}

export async function latestRevision(title: string): Promise<number> {
  const params = new URLSearchParams({
    action: "query", prop: "revisions", titles: title, rvprop: "ids", format: "json",
  });
  const resp = await fetch(`${API}?${params}`, { headers: { "User-Agent": USER_AGENT } });
  if (!resp.ok) throw new Error(`API ${resp.status} pour ${title}`);
  const pages = (await resp.json()).query.pages;
  const page = Object.values(pages)[0] as { revisions?: { revid: number }[] };
  if (!page.revisions) throw new Error(`page introuvable sur fr.wikipedia : ${title}`);
  return page.revisions[0].revid;
}

export async function fetchPage(title: string): Promise<PageSnapshot> {
  // Revision id BEFORE the HTML: an edit landing between the two requests
  // then leaves a stale stored revid, so the next watch run re-examines the
  // page instead of silently skipping the edit's polls.
  const revid = await latestRevision(title);
  const url = `https://fr.wikipedia.org/wiki/${title.replaceAll(" ", "_")}`;
  const resp = await fetch(url, { headers: { "User-Agent": USER_AGENT }, redirect: "follow" });
  if (!resp.ok) throw new Error(`fetch ${resp.status} pour ${url}`);
  // fetch() decodes strictly as UTF-8 here, which is exactly what we want —
  // httpx's charset sniffing once mis-decoded "août" and killed date parsing.
  return { title, revid, html: await resp.text() };
}

function parseDate(text: string): string | null {
  const cleaned = text.replace(/\[.*?\]/g, "").trim().toLowerCase().replaceAll(" ", " ");
  const m = cleaned.match(/(\d{1,2})\s+([a-zéûôà]+)\s+(\d{4})/);
  if (!m) return null;
  const month = MONTHS[m[2]];
  if (!month) return null;
  const dd = m[1].padStart(2, "0");
  return `${m[3]}-${String(month).padStart(2, "0")}-${dd}`;
}

/** Cells carry footnote markers, thin/no-break spaces, decimal commas and
 * en-dashes for "not asked". */
function parseNumber(text: string): number | null {
  const cleaned = String(text)
    .replace(/\[.*?\]/g, "")
    .replace(/[\s   ]/g, "")
    .replaceAll(",", ".")
    .replaceAll("%", "")
    .replaceAll("±", "");
  const m = cleaned.match(/-?\d+(?:\.\d+)?/);
  return m ? Number(m[0]) : null;
}

function normalizeFirm(text: string): string {
  let t = text.replace(/\[.*?\]/g, "").trim();
  t = t.replace(/\s*\/\s*La Presse$/, "");
  return FIRM_NORMALIZE[t] ?? t;
}

export function pollId(
  jurisdiction: string, region: string, firm: string, pollDate: string, sample: string | null,
): string {
  const key = `${jurisdiction}|${region}|${firm}|${pollDate}|${sample ?? ""}`;
  return createHash("sha1").update(key, "utf-8").digest("hex").slice(0, 16);
}

// The subgroup is never in the table; it is in the section heading above it.
const BREAKDOWN_PATTERNS = [
  "par langue", "par region", "par age", "par sexe", "par groupe",
  "francophone", "chez les", "ile de montreal", "region metropolitaine",
  "capitale-nationale", "reste du quebec",
];

const stripAccents = (text: string): string =>
  text.normalize("NFKD").replace(/[̀-ͯ]/g, "").replace(/[^\x00-\x7f]/g, "");

const isBreakdown = (headings: string[]): boolean => {
  const joined = stripAccents(headings.join(" / ")).toLowerCase();
  return BREAKDOWN_PATTERNS.some((p) => joined.includes(p));
};

const cleanHeading = (text: string): string =>
  text.replace("[modifier | modifier le code]", "").replace(/\s+/g, " ").trim();

/** (table, heading chain) in document order — querySelectorAll returns
 * document order, so this mirrors the Python xpath walk exactly. */
function* tablesWithHeadings(document: Document): Generator<[Element, string[]]> {
  const chain = new Map<number, string>();
  for (const node of document.querySelectorAll("h2,h3,h4,table")) {
    const tag = node.tagName.toLowerCase();
    if (tag !== "table") {
      const level = Number(tag[1]);
      chain.set(level, cleanHeading(node.textContent ?? ""));
      for (const k of [...chain.keys()]) if (k > level) chain.delete(k);
    } else {
      yield [node, [...chain.keys()].sort((a, b) => a - b).map((k) => chain.get(k)!)];
    }
  }
}

const cellLink = (cell: Element): string | null => {
  for (const a of cell.querySelectorAll("a")) {
    const href = a.getAttribute("href") ?? "";
    if (href.startsWith("http") && !href.includes("wikipedia.org") && !href.includes("wikimedia.org")) {
      return href;
    }
  }
  return null;
};

/** Rows that are not polls — the tables interleave narrative event rows like
 * "14 janvier 2026 : Annonce de la démission de François Legault" — are
 * dropped by requiring both a parseable date and >= 2 numeric party cells. */
export function parsePolls(
  snapshot: PageSnapshot, jurisdiction = "qc-provincial",
): { polls: PollRow[]; shares: ShareRow[] } {
  const { document } = parseHTML(snapshot.html);
  const pollRows: PollRow[] = [];
  const shareRows: ShareRow[] = [];
  const seen = new Set<string>();
  const skipped: [string, number][] = [];

  for (const [table, headings] of tablesWithHeadings(document)) {
    const firstTr = table.querySelector("tr");
    if (!firstTr) continue;
    const headers = [...firstTr.querySelectorAll("th")].map((th) => (th.textContent ?? "").trim());
    if (!headers.includes("Sondeur") || !headers.includes("CAQ")) continue;
    if (isBreakdown(headings)) {
      skipped.push([headings.at(-1) ?? "?", table.querySelectorAll("tr").length - 1]);
      continue;
    }
    const idx = new Map(headers.map((name, i) => [name, i]));
    const partyIdx = headers
      .map((h, i) => [h, i] as const)
      .filter(([h]) => h in PARTY_COLUMNS);

    for (const tr of table.querySelectorAll("tr")) {
      const cells = [...tr.children].filter((c) => c.tagName === "TD");
      if (cells.length < headers.length - 2) continue;

      const pollDate = parseDate(cells[0].textContent ?? "");
      if (pollDate === null) continue;

      const shares = new Map<string, number>();
      for (const [col, i] of partyIdx) {
        if (i >= cells.length) continue;
        const value = parseNumber(cells[i].textContent ?? "");
        if (value === null) continue;
        const code = PARTY_COLUMNS[col];
        shares.set(code, (shares.get(code) ?? 0) + value);
      }
      if (shares.size < 2) continue;

      const firmI = idx.get("Sondeur")!;
      if (firmI >= cells.length) continue;
      const firm = normalizeFirm(cells[firmI].textContent ?? "");
      if (!firm) continue;

      let sample: number | null = null;
      const sampleI = idx.get("Échantillon");
      if (sampleI !== undefined && sampleI < cells.length) {
        const raw = parseNumber(cells[sampleI].textContent ?? "");
        sample = raw !== null && raw > 0 ? Math.trunc(raw) : null;
      }

      let sourceUrl: string | null = null;
      const sourceI = idx.get("Source");
      if (sourceI !== undefined && sourceI < cells.length) sourceUrl = cellLink(cells[sourceI]);

      const pid = pollId(jurisdiction, "National", firm, pollDate, sample !== null ? String(sample) : null);
      if (seen.has(pid)) continue;
      seen.add(pid);

      pollRows.push({
        poll_id: pid,
        jurisdiction_code: jurisdiction,
        region_code: "National",
        firm,
        poll_date: pollDate,
        sample_size: sample,
        is_rolling: false,
        firm_rating: null,
        general_election: null,
        source_url: sourceUrl ?? `https://fr.wikipedia.org/wiki/${snapshot.title.replaceAll(" ", "_")}`,
      });
      for (const [code, pct] of shares) {
        shareRows.push({ poll_id: pid, party_code: code, pct_reported: pct });
      }
    }
  }

  if (skipped.length) {
    // Announce what was dropped: a silent filter looks identical to a source
    // that never had those tables.
    const detail = skipped.map(([name, n]) => `${name} (${n})`).join(", ");
    console.log(`    sous-groupes ignores : ${detail}`);
  }

  return { polls: pollRows, shares: shareRows };
}

/** All configured cycles, concatenated and de-duplicated on poll_id, plus the
 * revision id seen for each page so a watcher records what it processed. */
export async function fetchAll(
  cycles?: string[],
): Promise<{ polls: PollRow[]; shares: ShareRow[]; revisions: Record<string, number> }> {
  const polls: PollRow[] = [];
  const shares: ShareRow[] = [];
  const revisions: Record<string, number> = {};
  const seen = new Set<string>();
  for (const cycle of cycles ?? Object.keys(POLL_PAGES)) {
    const snapshot = await fetchPage(POLL_PAGES[cycle]);
    const parsed = parsePolls(snapshot);
    revisions[cycle] = snapshot.revid;
    for (const p of parsed.polls) {
      if (seen.has(p.poll_id)) continue;
      seen.add(p.poll_id);
      polls.push(p);
    }
    for (const s of parsed.shares) if (seen.has(s.poll_id)) shareIfNew(shares, s);
  }
  return { polls, shares, revisions };
}

const shareIfNew = (acc: ShareRow[], s: ShareRow): void => {
  // parsePolls already de-duplicates within a page; across pages a duplicate
  // poll_id keeps its first shares, matching polars' unique(keep="first").
  if (!acc.some((r) => r.poll_id === s.poll_id && r.party_code === s.party_code)) acc.push(s);
};
